arch/risc-v/src/mpfs/mpfs_ihc.c

```diff
@@ -980,6 +1091,24 @@ static int mpfs_rptun_stop(struct rptun_dev_s *dev)
 static int mpfs_rptun_notify(struct rptun_dev_s *dev, uint32_t notifyid)
 {
   uint32_t tx_msg[IHC_MAX_MESSAGE_SIZE];
+  uint32_t retries = 10000;
+  uint32_t flow_ctrl_en;
+  int      ret = OK;
+
+  /* INT_EN(0) is used by the bootloader (Linux) to indicate that
+   * it will be sending data.  Wait until it finishes first.
+   */
+
+  flow_ctrl_en = getreg32(MPFS_IHC_INT_EN(0));
```

Comment:
Is it possible that INT_EN is asserted right after this loop?

Why separate if first, instead of just direct while ((flow_ctrl_en != 0) && --retries) {} loop ?

Is it possible that linux hangs the whole nuttx here by never releasing the MPFS_IHC_INT_EN ?

Should there be some interrupt-wakeup based mutual exclusion between linux/nuttx instead? So that you wouldn't need to poll long times here?

Reply:
Yes, the loop could be optimized, but this loop takes only for a few microseconds at most; I have not seen a case where linux hangs nuttx (well, if it crashes it will do it so anyway).

The potential collision is later detected with:

      if (ctrl_reg & (ACK_INT | MP_MESSAGE_PRESENT))
        {
          spin_unlock_irqrestore(NULL, flags);
          return -EBUSY;
        }

Not sure it's 100% reliable.

 I wonder what kind of mutual exclusion could be done between Linux / nuttx, But the polling, when it happens, should not last long that one would notice. Of course, the opensbi nuttx could use a spinlock, that is also accessed via the hart2 NuttX, just the address would need to be fixed or something.

Reply:
I propose that we go with these at the moment and revisit after we have also Linux kernel update. What do you think?

Reply:
Probably a good idea, this has to be revisited anyway with more time. It could well be this sw flow control is no longer needed etc. (cannot test any time soon), but if needed, it could be reworked, but that takes some time with all the stress tests.